lib/Model/Importer/DebugInfo/DwarfImporter.cpp

```diff
@@ -451,83 +421,78 @@ class DwarfToModelConverter : public BinaryImporterHelper {
     return false;
   }
 
-  RecursiveCoroutine<const model::QualifiedType *>
-  getType(const DWARFDie &Die) {
-    if (auto MaybeType = Die.find(DW_AT_type)) {
-      if (MaybeType->getForm() == llvm::dwarf::DW_FORM_GNU_ref_alt) {
-        rc_return findAltType(MaybeType->getRawUValue());
+  RecursiveCoroutine<model::UpcastableType> makeType(const DWARFDie &Die) {
+    if (auto Type = Die.find(DW_AT_type)) {
+      if (Type->getForm() == llvm::dwarf::DW_FORM_GNU_ref_alt) {
+        rc_return Importer.findType({ AltIndex, Type->getRawUValue() }).copy();
       } else {
-        DWARFDie InnerDie = DICtx.getDIEForOffset(*MaybeType->getAsReference());
+        DWARFDie InnerDie = DICtx.getDIEForOffset(*Type->getAsReference());
         rc_return rc_recur resolveType(InnerDie, false);
       }
     } else if (auto MaybeOrigin = Die.find(DW_AT_abstract_origin)) {
       DWARFDie Origin = DICtx.getDIEForOffset(*MaybeOrigin->getAsReference());
-      rc_return rc_recur getType(Origin);
+      rc_return rc_recur makeType(Origin);
     } else {
       rc_return nullptr;
     }
   }
 
-  RecursiveCoroutine<model::QualifiedType> getTypeOrVoid(const DWARFDie &Die) {
-    using namespace model;
-    using PTK = model::PrimitiveKind::Values;
-    const model::QualifiedType *Result = rc_recur getType(Die);
-    if (Result != nullptr) {
-      rc_return *Result;
-    } else {
-      rc_return QualifiedType(Model->getPrimitiveType(PTK::Void, 0), {});
-    }
+  RecursiveCoroutine<model::UpcastableType>
+  makeTypeOrVoid(const DWARFDie &Die) {
+    if (model::UpcastableType Result = rc_recur makeType(Die))
+      rc_return Result;
+    else
+      rc_return model::PrimitiveType::makeVoid();
   }
 
-  RecursiveCoroutine<const model::QualifiedType *>
-  resolveTypeWithIdentity(const DWARFDie &Die,
-                          model::QualifiedType *DefinitionReference) {
+  RecursiveCoroutine<bool>
+  resolveTypeWithIdentity(const DWARFDie &Die, model::UpcastableType &Type) {
```

Comment:
Can you comment on the change of return type?

Reply:
Because the return value is never used. But it has to be a coroutine. IMO `bool` is less offensive than the pointer to the type.

Reply:
Ah uh, so coroutines cannot return `void`?

Reply:
Ah wait, it looks like there's one in PECOFF importer 🤔. My bad: yeah this should just be `void`.